### PR TITLE
Major speed improvement

### PR DESCRIPTION
**Pull request by JanTvrdik**
Significantly improves speed when validating large datasets.

https://github.com/stefk/JVal/pull/9 does improve speed as well.

[Before = master](https://blackfire.io/profiles/6cb54233-2cdd-4b7b-8499-e4db6b04a7e4/graph)
[After = #9 + #10](https://blackfire.io/profiles/84c3c4fc-fa83-4a91-83d3-ff6433420ccc/graph)

### CODE REVIEW

File: src/Constraint/PropertiesConstraint.php [hunk 1]
```diff
@@ -63,8 +63,37 @@ public function normalize(stdClass $schema, Context $context, Walker $walker)
...
+                $schemas[] = $schema->properties->{$property};
+            }
+
+            foreach ($schema->patternProperties as $regex => $propertySchema) {
+                if (Utils::matchesRegex($property, $regex)) {
+                    $schemas[] = $propertySchema;
+                }
+            }
+
+            if (!$schemas) {
```
Comment (stefk) [inline src/Constraint/PropertiesConstraint.php]: Scrutinizer is righty complaining about this. Please use `empty` or `count` instead of an implicit conversion.

File: src/Types.php [hunk 2]
```diff
@@ -26,6 +26,46 @@ class Types
     const TYPE_STRING = 'string';
 
     /**
+     * @var array
+     */
+    public static $phpToJson = [
+        'array' => self::TYPE_ARRAY,
+        'boolean' => self::TYPE_BOOLEAN,
+        'double' => self::TYPE_NUMBER,
+        'integer' => self::TYPE_INTEGER,
```
Comment (stefk) [inline src/Types.php:35]: Integers are also numbers as per the spec. Using a simple map like this one might cause issues with those types.

File: src/Types.php [hunk 3]
```diff
@@ -26,6 +26,46 @@ class Types
...
+    ];
+
+    /**
+     * @var array
+     */
+    public static $jsonToPhp = [
+        self::TYPE_ARRAY => 'array',
+        self::TYPE_BOOLEAN => 'boolean',
+        self::TYPE_INTEGER => 'integer',
+        self::TYPE_NUMBER => 'double',
```
Comment (stefk) [inline src/Types.php:48]: Same as above, numbers actually include integers.

File: src/Constraint/TypeConstraint.php [hunk 4]
```diff
@@ -80,21 +84,15 @@ public function normalize(stdClass $schema, Context $context, Walker $walker)
...
-            foreach ($schema->type as $type) {
-                if (Types::isA($instance, $type)) {
-                    $hasMatch = true;
-                    break;
-                }
-            }
 
-            if (!$hasMatch) {
+        } else {
+            if (!isset(Types::$phpToJson[$actualType]) || !in_array(Types::$phpToJson[$actualType], $schema->type)) {
```
Comment (stefk) [inline src/Constraint/TypeConstraint.php]: Here's an example of problems we could have with this mapping. Suppose you have:

``` php
$schema->type = ['string', 'number'];
$actualType = 'integer';
```

In that case, `in_array(Types::$phpToJson[$actualType], $schema->type)` will return false and a violation will be wrongly added to the context.

```
```

Comment (JanTvrdik) [inline src/Constraint/TypeConstraint.php]: This is currently handled in normalize phase, see https://github.com/stefk/JVal/pull/10/commits/b2f7fd7dcc4903769429fa70d88ea1512ecfaa12#diff-96dad29cd96ce01c1dbbd358fd2594e5R65.

File: src/Types.php [hunk 5]
```diff
@@ -36,58 +76,25 @@ class Types
...
-
-    /**
-     * Returns whether an instance matches a given type.
-     *
-     * @param mixed  $instance
-     * @param string $type
-     *
-     * @return bool
-     */
-    public static function isA($instance, $type)
```
Comment (stefk) [inline src/Types.php:67]: What about restoring this method and make it use the `$phpJsonCompatibility` map internally? Would that hurt performance?

**Comment by JanTvrdik** [comment]
Generally the TypeConstraint optimizations were the hardest to do. I must have rewrote the entire code at least three times. The JSON type system is almost the same as PHP type system, but not quite, which makes it very difficult to write the code in way that is both fast and easy read.

Even we the carefully crafted optimizations, the TypeConstraint is still visible on profiler as one of the slowest part because it's called so many times.

File: src/Types.php (hunk 2 above)
Comment (JanTvrdik) [inline src/Types.php:35]: Those arrays are there mostly to replace the slow switch statements (with linear complexity) with hash lookups (with constant complexity).

File: src/Constraint/TypeConstraint.php (hunk 4 above)
Comment (stefk) [inline src/Constraint/TypeConstraint.php]: Right, I forgot about that.

File: src/Types.php (hunk 5 above)
Comment (JanTvrdik) [inline src/Types.php:67]: Not by a lot, but it is measurable. On my data instance it makes the whole schema validation about 4 % slower (500 ms instead of 480ms).

File: src/Types.php (hunk 2 above)
Comment (stefk) [inline src/Types.php:35]: I understand. That's a good idea. I'm just concerned about exposing those maps with the number/integer trick. This might be a source of bugs in the long run.

**Comment by JanTvrdik** [comment]
I added [another optimization for Context](https://github.com/stefk/JVal/pull/10/commits/d4a252d96cd2c528a0df9e52bfcbf0ba9d48475c)

File: src/Context.php [hunk 6]
```diff
@@ -60,11 +64,7 @@ public function enterSibling($pathSegment)
      */
     public function leaveNode()
     {
-        if (count($this->pathSegments) === 0) {
-            throw new \LogicException('Cannot leave node: instance stack is empty');
-        }
```
Comment (stefk) [inline src/Context.php:65]: I think this should stay. It's a basic check, and it's almost free with your implementation
(`$this->pathLength === 0`).

File: src/Context.php [hunk 7]
```diff
@@ -74,7 +74,8 @@ public function leaveNode()
      */
     public function getCurrentPath()
     {
-        return $this->pathSegments ? '/'.implode('/', $this->pathSegments) : '';
+        $this->pathSegments = array_slice($this->pathSegments, 0, $this->pathLength);
```
Comment (stefk) [inline src/Context.php:81]: Isn't this a no-op?

Comment (stefk) [inline src/Context.php:81]: Nevermind, I got it.

**Comment by stefk** [comment]
@JanTvrdik, thanks a lot for your work on this one. It's obviously going to be merged but there's two remaining points I'd like to see addressed before that:
1. I notice that the current/master implementation of `Types` and `TypeConstraint` produce redundant calls to `gettype`. I'd like to know what would be the performance trade-off of simply moving the `gettype` call out of `isA` (thus making it: `isA($phpType, $primitiveType)`) _plus_ using your internal data structure (`$phpJsonCompatibility`). It's almost what you did, except there'd be no need to leak tricky type mapping stuff in the codebase. If that improvement is really not sufficient, at least a basic doc with warnings about gotchas should be added wherever necessary.
2. As is, your optimization of the walker is quite hard to follow (especially with properties like `cacheA` and `cacheB`). As filtering and caching constraints by version and type (what you call `cacheB`) isn't really tied to a particular schema, it would be best IMO to move it to the registry. For example, `getConstraints` could accept a second parameter `$typeFilter`for that purpose. The registry would handle a cache by version (the current `$constraints`) plus a cache by version and type. The walker on the other hand would deal only with one cache by schema and type.

**Comment by JanTvrdik** [comment]
Both the Walker and TypeConstraint should now be more readable.

**Comment by stefk** [comment]
That's great, thanks again!
